File: java/NetworkRouting/src/main/java/com/yogesh/util/RoutingHelper.java
package com.yogesh.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Map.Entry;
import java.util.PriorityQueue;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import jdk.nashorn.internal.runtime.regexp.RegExp;

import org.springframework.stereotype.Component;

import com.yogesh.routing.NetworkMap;
import com.yogesh.routing.Route;
import com.yogesh.routing.Station;

/**
 * @author Yogesh Sharma
 * 
 *         This is the helper class initialize networkmap and find distance and
 *         other helper methods
 * 
 */
@Component("routingHelper")
public class RoutingHelper {

	/**
	 * this method to initialize network take input from stdin and add routes to
	 * network map
	 * 
	 * @param inputs
	 *            - List string which represent route like
	 *            <source>-><destination>:traveltime
	 * @param networkMap
	 *            - base network map
	 * @throws Exception
	 *             - throw exceptions if any input validation fails or other
	 */
	public void initNetworkMap(List<String> inputs, NetworkMap networkMap)
			throws Exception {
		String source = null;
		String destination = null;
		double travelTime = 0;
		String ttStr = null;

		for (String input : inputs) {
			try {
				source = input.substring(0, input.indexOf("->"));
				if (!isValidStaion(source)) {
					throw new Exception(
							source
									+ " -- is invalid source, special characters not allowed");
				}
				destination = input.substring(input.indexOf("->") + 2,
						input.indexOf(":"));
				if (!isValidStaion(destination)) {
					throw new Exception(
							destination
									+ " -- is invalid destination, special characters not allowed");
				}
				ttStr = input.substring(input.indexOf(":") + 1, input.length());
				travelTime = Double.valueOf(ttStr);
				if (travelTime < 0) {
					throw new Exception(
							ttStr
									+ " -- is invalid travel time, travel time must be +ve number");
				}

			} catch (NumberFormatException ne) {
				throw new Exception(ttStr + " -- is invalid travel time");
			}
			networkMap.addRoute(source, destination, travelTime);
		}
	}

	private boolean isValidStaion(String station) {
		Pattern p = Pattern.compile("[a-zA-Z0-9]");
		return p.matcher(station).matches();
	}

	static NetworkMap findAllBestRouteForSource(Station source) {
		NetworkMap map = new NetworkMap();
		// Algo:
		// 1. Take the unvisited node with minimum weight.
		// 2. Visit all its neighbours.
		// 3. Update the distances for all the neighbours (In the Priority
		// Queue).
		// Repeat the process till all the connected nodes are visited.

		source.minDistance = 0;
		PriorityQueue<Station> queue = new PriorityQueue<Station>();
		queue.add(source);

		while (!queue.isEmpty()) {

			Station u = queue.poll();

			for (Route neighbour : u.neighbours) {
				Double newDist = u.minDistance + neighbour.travelTime;

				if (neighbour.target.minDistance > newDist) {
					// Remove the node from the queue to update the distance
					// value.
					queue.remove(neighbour.target);
					neighbour.target.minDistance = newDist;

					// Take the path visited till now and add the new node.s
					neighbour.target.path = new LinkedList<Station>(u.path);
					neighbour.target.path.add(u);

					// Reenter the node with new distance.
					queue.add(neighbour.target);
				}
			}
		}
		map.getStations().put(source.name, source);
		return map;

	}

	/**
	 * @param networkMap
	 * @param travelTime
	 */
	static List<Station> printNearByStations(NetworkMap networkMap,
			double travelTime) {
		List<Station> nearByStations = new ArrayList<Station>();
		if (networkMap.getStations() == null
				|| networkMap.getStations().isEmpty()) {
			System.out.println("Empty network map...");
		}
		List<Station> Stations = new LinkedList<Station>(networkMap
				.getStations().values());
		// Sort all station to have near one 1st
		Collections.sort(Stations);
		StringBuffer sb = new StringBuffer();
		for (Station station : Stations) {
			if (station.minDistance > 0 && station.minDistance <= travelTime) {
				sb.append(station + ": " + station.minDistance + ", ");
				nearByStations.add(station);
			}
		}
		System.out.println(sb.substring(0, sb.length() - 2));
		return nearByStations;
	}

	/**
	 * @param networkMap
	 * @param destination
	 */
	static double printRoute(NetworkMap networkMap, String destination) {
		if (networkMap.getStations() == null
				|| networkMap.getStations().isEmpty()) {
			System.out.println("Empty network map...");
		}
		for (Entry<String, Station> entry : networkMap.getStations().entrySet()) {
			if (entry.getValue().name.equals(destination)) {
				LinkedList<Station> path = entry.getValue().path;
				for (Station station : path) {
					System.out.print(station + " -> ");
				}
				System.out.print(entry.getValue() + ": "
						+ entry.getValue().minDistance + "\n");
				return entry.getValue().minDistance;
			}
		}

		return -1;
	}

	static void printNetworkMap(NetworkMap networkMap) {
		if (networkMap.getStations() == null
				|| networkMap.getStations().isEmpty()) {
			System.out.println("Empty network map...");
		}
		for (Entry<String, Station> s : networkMap.getStations().entrySet()) {
			LinkedList<Station> path = s.getValue().path;
			for (Station station : path) {
				System.out.print(station + " -> ");
			}
			System.out.print(s.getValue() + ": " + s.getValue().minDistance
					+ "\n");
		}
	}

}

File: scala/src/main/scala/com/yogesh/routing/util/RoutingHelper.scala
package com.yogesh.routing.util

import com.yogesh.routing.NetworkMap
import java.util.regex.Pattern
import com.yogesh.routing.Station

import java.util.PriorityQueue
import scala.collection.mutable.LinkedList
import scala.collection.immutable.ListMap
import scala.collection.mutable.MutableList

class RoutingHelper {

  // this can be implemented to cache results for searched results for given source
  var mapCache = collection.mutable.Map[String, NetworkMap]()

  /**
   * method is initialize networkmap with given input routes
   * @param input
   * @param networkMap
   */
  def initNetworkMap(input: List[String], networkMap: NetworkMap): Unit = {

    if (input.size < 1) {
      throw new Exception("input routes list can not be empty")
    }

    for (inputStr <- input) {
      try {
        val source = inputStr.substring(0, inputStr.indexOf("->"))
        isValidStaion(source)
        val destination = inputStr.substring(inputStr.indexOf("->") + 2, inputStr.indexOf(":"))
        isValidStaion(destination)
        val ttStr = inputStr.substring(inputStr.indexOf(":") + 1, inputStr.length())
        val travelTime = ttStr.toDouble
        if (travelTime < 0) {
          throw new Exception("travel time must be +ve number")
        }
        networkMap.addRoute(source, destination, travelTime)
      } catch {
        case t: Throwable =>
          t.printStackTrace()
          throw t;
      }

    }

  }

  /**
   * validation for not include special characters
   * @param station
   */
  def isValidStaion(station: String): Unit = {
    if (!Pattern.compile("[a-zA-Z0-9]").matcher(station).matches()) {
      throw new Exception(station + "is not valid")
    }
  }

  def printNetworkMap(networkMap: NetworkMap): Unit = {

    if (networkMap.stations == null || networkMap.stations.isEmpty) {
      print("Empty network map...")
    }
    print("\n")
    networkMap.stations foreach (x => {
      x._2.path foreach { p => println(" " + p.name + "->") }
      println(x._2.name + x._2.minDistance)
    })
  }

  /**
   * method to find best shortest path for given source
   *
   * @param source - Source station
   */
  def findAllBestRouteForSource(source: Station): Unit = {

    source.minDistance = 0

    var queue = new PriorityQueue[Station]()

    queue.add(source)

    while (!queue.isEmpty) {

      val u = queue.poll

      u.neighbours foreach { neighbour =>
        {
          val newDist = u.minDistance + neighbour.travelTime

          if (neighbour.target.minDistance > newDist) {
            queue remove neighbour.target
            neighbour.target.minDistance = newDist
            // Take the path visited till now and add the new node.s
            neighbour.target.path += u
            // Reenter the node with new distance.
            queue.add(neighbour.target)
          }

        }
      }
    }
  }

  /**
   * method to print shortest path for given destination station
   * @param networkMap - network map with calculated shortest path from source
   * @param destination - destination station where want to go
   * @return  - time travel if find path OR -1 if no path found
   */
  def printRoute(networkMap: NetworkMap, destination: String): Double = {
    var retVal = -1.0
    if (networkMap.stations == null || networkMap.stations.isEmpty) {
      print("Empty network map...")
    }
    // filter network map for given destination
    val filterMap = networkMap.stations.filterKeys { _ == destination }

    filterMap foreach (x => {
      x._2.path foreach { p => { print(" " + p.name + "->") } }
      print(x._2.name + ":" + x._2.minDistance)
      retVal = x._2.minDistance
    })

    retVal
  }

  /**
   * print nearby station for source and given time tavel limit
   *
   * @param networkMap - networkMap with shorted route calculated
   * @param travelTime - time travel limit
   * @return - list of stations where we can reach in given time travel
   */
  def printNearByStations(networkMap: NetworkMap, travelTime: Double): MutableList[Station] = {
    if (networkMap.stations == null || networkMap.stations.isEmpty) {
      print("Empty network map...")
    }

    val filterMap = networkMap.stations.filter { case (k, v) => v.minDistance <= travelTime && v.minDistance != 0 }
    var stationList = new MutableList[Station]()

    if (!filterMap.isEmpty) {
      val sortedMap = ListMap(filterMap.toSeq.sortBy(_._2): _*)

      var allNearby = new java.lang.StringBuilder

      sortedMap foreach { item =>
        {
          allNearby.append(item._2.name + ":" + item._2.minDistance + ",")
          stationList.+=(item._2)
        }
      }
      println("\n" + allNearby.toString().substring(0, allNearby.toString().length() - 1))
    }
    stationList
  }

}

object RoutingHelper {
  implicit def createRoutingHelper(): RoutingHelper = {
    new RoutingHelper()
  }
}